miio/airhumidifier_jsq.py

```diff
@@ -0,0 +1,330 @@
+        "power",  # (0: off, 1: on)
+        "no_water",  # (0: enough, 1: add water)
+        "lid_opened",  # (0: ok, 1: lid is opened)
+    ]
+}
+
+
+class OperationMode(enum.Enum):
+    Intelligent = 0
+    Level1 = 1
+    Level2 = 2
+    Level3 = 3
+    Level4 = 4
+
+
+class LedBrightness(enum.Enum):
+    Off = 0
+    Low = 1
+    High = 2
+
+
+class AirHumidifierStatus:
+    """Container for status reports from the air humidifier jsq."""
+
+    def __init__(self, data: Dict[str, Any], device_info: DeviceInfo) -> None:
+        """
+        Status of an Air Humidifier (shuii.humidifier.jsq001):
+            [24, 30, 1, 1, 0, 2, 0, 0, 0]
+
+        Parsed by AirHumidifierJsq device as:
+            {'temperature': 24, 'humidity': 29, 'mode': 1, 'buzzer': 1,
+             'child_lock': 0, 'led_brightness': 2, 'power': 0, 'no_water': 0,
+              'lid_opened': 0}
+        """
+        self.data = data
+        self.device_info = device_info
+
+    @property
+    def power(self) -> str:
+        """Power state."""
+        return "on" if self.data["power"] == 1 else "off"
+
+    @property
+    def is_on(self) -> bool:
+        """True if device is turned on."""
+        return self.power == "on"
+
+    @property
+    def mode(self) -> OperationMode:
+        """Operation mode. Can be either low, medium, high or humidity."""
+
+        try:
+            mode = OperationMode(self.data["mode"])
+        except ValueError as e:
+            _LOGGER.exception("Cannot parse mode: %s", e)
+            return OperationMode.Intelligent
+
+        return mode
+
+    @property
+    def temperature(self) -> int:
+        """Current temperature in degree celsius."""
+        return self.data["temperature"]
+
+    @property
+    def humidity(self) -> int:
+        """Current humidity in percent."""
+        return self.data["humidity"]
+
+    @property
+    def buzzer(self) -> bool:
+        """True if buzzer is turned on."""
+        return self.data["buzzer"] == 1
+
+    @property
+    def led_brightness(self) -> LedBrightness:
+        """Buttons illumination Brightness level."""
+        try:
+            brightness = LedBrightness(self.data["led_brightness"])
+        except ValueError as e:
+            _LOGGER.exception("Cannot parse brightness: %s", e)
+            return LedBrightness.Off
+
+        return brightness
+
+    @property
+    def child_lock(self) -> bool:
+        """Return True if child lock is on."""
+        return self.data["child_lock"] == 1
+
+    @property
+    def no_water(self) -> bool:
+        """True if the water tank is empty."""
+        return self.data["no_water"] == 1
+
+    @property
+    def lid_opened(self) -> bool:
+        """True if the water tank is detached."""
+        return self.data["lid_opened"] == 1
+
+    @property
+    def firmware_version(self) -> str:
+        """Returns the fw_ver of miIO.info. For example 1.3.9."""
+        return self.device_info.firmware_version
```

Comment:
Better skip properties accessing the `info` here. They are unnecessary and easily accessible through the common info interface for the use cases that require that information.

Reply:
Ok, so just remove `hardware_version` and `firmware_version` methods from `AirHumidifierStatus`? Impl is based on `airhumidifier.py` which defines a similar method. Maybe you need to do cleanup there also.

Reply:
Yes, I would say so.

Btw, is the device so much different than adding support for this device into the `airhumidified.py` would have unnecessarily complicated that?

Reply:
`airhumidified.py` isn't a general minimal humidifier impl, it has options with not available for this one, e.g:  `target_humidity`, `motor speed`, `dry`, `depth`, `use_time`, `button_pressed`

Also `airhumidified.py` uses different `mode` and they are named string values, this humidifier uses int values. Brightness modes are different,  e.g. OFF state is 2 for  `airhumidified.py`  and 0 for this one. Status read/write is different, here we get all values at once, and in  `airhumidified.py` we fetch single attribute by name if I'm not mistaken.

> would have unnecessarily complicated that?

I don't think so, but it will be good to revise API and make one that covers all existing use cases.

Reply:
Ok, sounds good to me! Is there something else you want to change before getting this merged?

Reply:
Yes, I'll push firmware/hardware related cleanup soon + also ensure that everything still works in my Home Assistant with updated support of the humidifier in the corresponding HASS component

Reply:
Done